x/ibc/02-client/keeper/keeper.go

```diff
@@ -177,20 +177,29 @@ func (k Keeper) GetLatestClientConsensusState(ctx sdk.Context, clientID string)
 
 // GetClientConsensusStateLTE will get the latest ConsensusState of a particular client at the latest height
 // less than or equal to the given height
-func (k Keeper) GetClientConsensusStateLTE(ctx sdk.Context, clientID string, maxHeight uint64) (exported.ConsensusState, bool) {
-	for i := maxHeight; i > 0; i-- {
-		found := k.HasClientConsensusState(ctx, clientID, i)
+func (k Keeper) GetClientConsensusStateLTE(ctx sdk.Context, clientID string, maxHeight exported.Height) (exported.ConsensusState, bool) {
+	// NOTE: For tendermint Heights this is only implemented for a single epoch
+	height := maxHeight
+	var success bool
+	for height.Valid() {
+		found := k.HasClientConsensusState(ctx, clientID, height)
 		if found {
-			return k.GetClientConsensusState(ctx, clientID, i)
+			return k.GetClientConsensusState(ctx, clientID, height)
+		}
+		height, success = height.Decrement()
+		if !success {
+			break
 		}
 	}
 	return nil, false
 }
 
 // GetSelfConsensusState introspects the (self) past historical info at a given height
 // and returns the expected consensus state at that height.
-func (k Keeper) GetSelfConsensusState(ctx sdk.Context, height uint64) (exported.ConsensusState, bool) {
-	histInfo, found := k.stakingKeeper.GetHistoricalInfo(ctx, int64(height))
+func (k Keeper) GetSelfConsensusState(ctx sdk.Context, height exported.Height) (exported.ConsensusState, bool) {
+	// Only support retrieving HistoricalInfo from the current epoch for now
+	// TODO: check EpochNumber matches expected
```

Comment:
This should work across epochs, I think, so that handshakes will work cleanly across upgrades

(we should use both the epoch number & the epoch height in the key)